exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/efgs/OperationDao.java

```diff
@@ -1,161 +1,127 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
 import static java.util.Objects.requireNonNull;
 
 @Repository
 public class OperationDao {
+    public enum EfgsOperationState {STARTED, FINISHED, ERROR}
 
-    private static final Logger LOG = LoggerFactory.getLogger(OperationDao.class);
-
-    public enum EfgsOperationState {QUEUED, STARTED, FINISHED, ERROR}
     public enum EfgsOperationDirection {INBOUND, OUTBOUND}
-    private static final long MAX_RUN_COUNT = 2;
+
+    public static final long STALLED_MIN_AGE_IN_MINUTES = 10;
 
     private final NamedParameterJdbcTemplate jdbcTemplate;
 
     public OperationDao(NamedParameterJdbcTemplate jdbcTemplate) {
         this.jdbcTemplate = requireNonNull(jdbcTemplate);
     }
 
-    public long getQueueId() {
-        String sql = "select id from en.efgs_operation where state = cast(:state as en.state_t) and " +
-                "direction = cast(:direction as en.direction_t)";
-        return requireNonNull(jdbcTemplate.queryForObject(sql, Map.of(
-                "state", OperationDao.EfgsOperationState.QUEUED.name(),
-                "direction", OperationDao.EfgsOperationDirection.OUTBOUND.name()
-        ), Long.class));
-    }
-
-    @Transactional(timeout = 10)
-    public Optional<Long> startOutboundOperation() {
-        String lock = "lock table en.efgs_operation in access exclusive mode";
-        jdbcTemplate.update(lock, Map.of());
-        if (isOutboundOperationAvailable()) {
-            long operationId = this.getQueueId();
-            markOutboundOperationStartedAndCreateNewQueue(operationId);
-            return Optional.of(operationId);
-        } else {
-            LOG.info("Update to efgs unavailble. Skipping.");
-            return Optional.empty();
-        }
-    }
-
-    public boolean finishOperation(long operationId, int keysCountTotal) {
-        String sql = "update en.efgs_operation set state = cast(:new_state as en.state_t), keys_count_total = :keys_count_total, " +
-                "run_count = run_count + 1 " +
+    public boolean finishOperation(long operationId, int keysCountTotal, Optional<String> batchTag) {
+        String sql = "update en.efgs_operation set state = cast(:new_state as en.state_t), batch_tag = :batch_tag, " +
+                "keys_count_total = :keys_count_total, updated_at = :updated_at " +
                 "where id = :id";
-        return jdbcTemplate.update(sql, Map.of(
-                "new_state", EfgsOperationState.FINISHED.name(),
-                "id", operationId,
-                "keys_count_total", keysCountTotal
-        )) == 1;
+        Map<String, Object> params = new HashMap<>();
+        params.put("new_state", EfgsOperationState.FINISHED.name());
+        params.put("id", operationId);
+        params.put("batch_tag", batchTag.orElse(null));
+        params.put("keys_count_total", keysCountTotal);
+        params.put("updated_at", new Timestamp(Instant.now().toEpochMilli()));
```

Comment:
This could also be generated in the db

Reply:
Altthough, as I read further, it's clear that it's compared with Java time later on. So no point moving this in to DB time, unless all of that is moved as well.

Reply:
Yep, this ia a little bit nasty part. It's still shaping, I think, let's see, but for now I let it be.